### ISSUE TITLE
[FEAT] React-Query v5 compatibility

### ISSUE THREAD

**Issue by bombillazo** [refinedev/refine#5370]
### Is your feature request related to a problem? Please describe.

We've created a branch to migrate our application code to RQ v5, but have refrained since Refine uses RQ under the hood and want to avoid breaking any dependencies.

### Describe alternatives you've considered

Keep using RQ v4 for the time being.

### Additional context

https://tanstack.com/query/v5/docs/react/guides/migrating-to-v5

The migration process is relatively straightforward. They did remove the query callbacks, and the new paradigm for reacting to queries is doing so with `useEffect` hooks. One can still set global callbacks for queries, which refine uses for the notifications context, so the migration for Refine will likely center around that for their custom useQuery hooks.

They've also standardized the query function signatures only to accept objects. The typing should auto-alert the internal refine code.

### Describe the thing to improve

Having the ability to use React Query v5 in our refine project.

**Comment by BatuhanW** [comment]
Thanks for the issue @bombillazo.

Migration looks straightforward, other than callbacks. @aliemir had the idea of providing these callbacks as a polyfill with the current v4 version of refine.

**Comment by bombillazo** [comment]
Yeah! I think that would be the best approach for backward compatibility

**Comment by rilrom** [comment]
Would migrating to v5 allow us to take advantage of the react query suspense features?

**Comment by zernie** [comment]
yea suspense would be great

**Comment by bombillazo** [comment]
any update on this?

**Comment by zernie** [comment]
@BatuhanW could you please specify what needs to be done in preparation before migration to react-query v5?

**Comment by BatuhanW** [comment]
Hey @zernie the main refactor should be replacing removed onXXX calbacks with useEffect for notifications in internal logic. Also, since these callbacks might be used by current Refine users, we also need to keep these types and trigger them with useEffect. Other thing is we need to check react-query's new types and make sure there is no breaking change with current Refine types.

We are open to contributions for this issue, if anyone wants to work on this.

**Comment by tuffstuff9** [comment]
+1 for adopting v5!

**Comment by bombillazo** [comment]
Was anyone able to look into this?  The Refine source is pretty daunting and internally has lots of file/usage of RQ, perhaps identifying _where_ the changes are required will motivate others to take the challenge?

**Comment by kziemski** [comment]
Curious if this is on roadmap and/or in need of contributors? Haven't seen any mention of it in awhile.

**Comment by BatuhanW** [comment]
Hey @kziemski we are open to contributions for this one.

**Comment by kziemski** [comment]
@BatuhanW   apologies wasn't expecting such a quick response(s). is agnostic the next major version release or is that not yet determined?

**Comment by zernie** [comment]
@BatuhanW wait, does that mean it'd be possible to use RTK Query in the future?

**Comment by BatuhanW** [comment]
We don't want to release this update as a basic version bump, but we want to deeply integrate new features such as suspense, server components into Refine. Things are moving fast nowadays on RSC side. We'll keep you posted in the future.

**Comment by kziemski** [comment]
And another library becomes infected with RSC. Really was hoping to keep this library as part of my stack.  oh well good luck

**Comment by BatuhanW** [comment]
> And another library becomes infected with RSC. Really was hoping to keep this library as part of my stack. oh well good luck

We don't want to enforce anything on users. That's why we are not doing the upgrade at the moment. Even if we _support_ some features, that wouldn't be required or would be opt-in. RSC is controversial topic right now, we'll wait for it to become more mature and see how community positions itself with it.

**Comment by kziemski** [comment]
I think you're right to hold back. React core's head space is clearly in supporting vercel/nextjs priorities.  
Thankfully, this time they tripped over tanner's library. mess with the wrong library and you will incur the wrath of x.

I was just hoping to keep the libraries version in sync as react-query v5 is a dependency but i'll make do. ty for the response.

**Comment by datner** [comment]
Hey, this makes new projects that use pnpm start out broken because of how pnpm treats peers. I'm very interested in seeing this resolved, even if not properly solved, as the current situation is a proper blocker on the whole thing..
I am willing to help if it's a matter of effort, I have some familiarity with the project internals

**Comment by zernie** [comment]
>  we also have plans to make react-query

that'd be perfect since we use redux. is there any way to contribute to this effort?

**Comment by zernie** [comment]
Forgot to mention @BatuhanW

**Comment by BatuhanW** [comment]
Hey @zernie sorry, your comment is overlooked among bloated GitHub notifications, my colleague @aliemir will provide some context and our plans on this issue.

**Comment by aliemir** [comment]
Hey everyone 👋

I'd like to share some context on the delay in updating Refine to use TanStack React Query v5 and our plans moving forward.

The main reason for the delay is our focus on ensuring the new features in Refine v4 are stable and won't require breaking changes later on. Given that our last major release included updates across almost all providers, we wanted to be cautious and make sure everything is mature enough before introducing another major change.

We also wanted to make sure that we're in a stable environment for new features such as Server components etc. Our approach is to release the upcoming improvements in Refine gradually, allowing users to adapt their codebases smoothly. Once these changes are stable, we'll proceed with updating React Query to the latest version, which will be part of a major release from `@refinedev/core`.

The migration from React Query v4 to v5 is relatively straightforward for Refine; it mainly involves renaming a few props and updating internal query callbacks. However, for users who use `@tanstack/react-query` separately or rely heavily on deprecated options from `queryOptions` and `mutationOptions`, some manual adjustments will be needed. To make this transition as easy as possible, we'll roll out the upcoming Refine improvements first, so when the major release occurs, users will only need to focus on adjusting for the React Query changes in their Refine projects.

We'll keep this issue updated. You can expect to hear from us soon with a roadmap of the improvements, new features and eventually the major update.

**Comment by datner** [comment]
This means that Refine v4 would have a non-major breaking version update? Maybe refine should adopt the epoch approach as outlined by the semver spec creator and github cofounder [here](https://tom.preston-werner.com/2022/05/23/major-version-numbers-are-not-sacred.html)? It feels like the major increment is more of a hinderance than a blessing for refine, and I can totally relate

**Comment by sebbean** [comment]
snaaps i'm in monorepo deps hell, cant feasible downgrade to react-query 4, mayyy be down to help upgrade to 5 if its simple enough

**Comment by bombillazo** [comment]
@sebbean I saw the [migration changes](https://tanstack.com/query/latest/docs/framework/react/guides/migrating-to-v5) from v4 to v5, mostly some renames and straight forward function signature changes. The biggest change was the removal of the useQuery callbacks, Im not sure how of those many Refine is using.

**Comment by taniaelysia42** [comment]
Hi Guys, any update to upgrade react-query into v5? because i also have problem with monorepo deps

**Comment by qstyler** [comment]
Also would like to hear an update.

**Comment by daxaxelrod** [comment]
Same here, our backend is trpc v5 and mutation issued from refine always throw
`Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'mutation')`

**Comment by carakessler** [comment]
@aliemir any update? a timeline would be super helpful here

**Comment by mi5ha** [comment]
@aliemir This is becoming a really big problem, any roadmap on this? 😊

**Comment by alicanerdurmaz** [comment]
We’re releasing Refine AI in mid-June. Right after that, we’ll review the React Query v5 upgrade along with other key priorities like React 19 and start planning the implementation.

Apologies for the delay and thank you for your patience. We’ll prepare and share the roadmap with you as soon as possible.
